Components and Matrix Operations for CellInfo section on MC3D

Bud-Macaulay opened the pull request:
Moderately sizable commit
 - Majority of changes are for the 3 different switches (Wrapped Button componts) for use throughout MC.
- Lattice params are rendered on both MC2D and MC3D added matrixUtils and formulaUtils to be used in calculating properties (conventional Cell and Density)
- I've decided to not add the periodic_table script since this was a one time utility and it's unclear whether we'd need this tool.

Bud-Macaulay committed:
Added Wrapped Buttons (FlipperSwitch, ToggleSwitch) UI components. Moved Formula Calcs that were on MC3D so they are availible in MC2D etc. Made MatrixUtils with common matrix utilities (predominantly used in Lattice Conversion at the moment)

Bud-Macaulay commented:
Quick note - in docs it mentions updating the version of this library when making changes - wasnt sure whether to do that before the PR or after.

Bud-Macaulay committed:
renamed jsx to js, moved PrimToConv into symmetry

eimrek reviewed (changes requested):
Thanks @Bud-Macaulay. Im getting an error when trying to open the demo page:

```
Failed to resolve import "./utils/matrixUtils.jsx" from "lib/main.js". Does the file exist?
```

Bud-Macaulay committed:
Fix broken links in main.js

eimrek commented on lib/utils/formulaUtils.jsx:
```diff
@@ -0,0 +1,40 @@
+import { formatChemicalFormula } from "./formatting";
+
+export function countNumberOfAtoms(formula) {
```
Could you here write explicitly that hillFormula is needed?

eimrek commented on lib/utils/formulaUtils.jsx:
```diff
@@ -0,0 +1,40 @@
+  return num;
+}
+
+export function countNumberOfElements(formula) {
```
Maybe also here hillFormula?

eimrek commented on lib/utils/formulaUtils.jsx:
```diff
@@ -0,0 +1,40 @@
+  return formula.split(/(?=[A-Z])/).length;
+}
+
+export function formatTitle(formulaStr, id, method) {
```
This one, I think, is not very generally usable, and quite specific to the MC3D frontend. I would keep it there.

eimrek commented on lib/utils/formulaUtils.jsx:
```diff
@@ -0,0 +1,40 @@
+  );
+}
+
+export function calculateDensity(formula, pt, volume) {
```
Could you include the periodic table in this repo, and remove the argument?

eimrek commented on lib/utils/matrixUtils.js:
```diff
@@ -0,0 +1,136 @@
+  return result;
+}
+
+export function getMatrixParams(matrix) {
```
This could be made more specific to be clear it's to get the cell representation in lengths + angles. Could the name and/or comments/doc be improved?

eimrek commented on src/App.jsx:
```diff
@@ -50,13 +53,44 @@ function App() {
+          fontSize="10px"
         />
       </div>
+      <div style={{ padding: "8px" }}>
```
Here, padding will mess up a bit the red border, that is supposed to show the exact borders of the component. Use margin here.

eimrek commented on src/App.jsx:
```diff
@@ -50,13 +53,44 @@ function App() {
+        <ToggleSwitch switchLength="120px" />
+      </div>
+      <span>FlipperSwitch</span>
+      <div style={{ padding: "4px" }}>
```
The labels for this switch have a fairly unintuitive location, which could be improved. Or if we're not really using this switch, it could also be removed.

eimrek reviewed (changes requested):
Looks good, but left some comments for some changes.

Also, would be great to have some basic `console.log` calls in the demo page for some of the utility functions (just for a very basic check that everything is working).

Additionally, for the switches, it would be great to use css modules (see the other components), just to be consistent.

And finally, for matrixUtils / formulaUtils, maybe the name could not include Utils, to be consistent with the other files in the folder.

Bud-Macaulay committed:
Making requested changes/added console matrix Tests

Bud-Macaulay commented:
All changes done.

- Cleaned up the src, removing ToggleSwitchBig and the FlipperSwitch as they are currently not used anywhere on MCXD
- renamed these matrixUtils.js and formulaUtils.jsx to matrix.js and formula.js [ if thats okay]
- hillFormula in prop for formula.js [perhaps in free time i can write some tests to see how this could break/easy fixed]
- updated components/ToggleSwitch/ to use styles.module.css.

- I have not made changes to ToggleSwitchBig and FlipperSwitch, if okay to leave them as is until we find a use case for them?  

- I've added a js function in a new directory called 'tests' that console logs matrix operations on a test matrix. For now this seems reasonable?

eimrek reviewed (changes requested):
Thanks, looks great. Some final things:

* if you want to keep ToggleSwitchBig and FlipperSwitch, also keep them in the demo page. To see what they look like, and how to use them. Otherwise the code will be invisible.
* The simpleTests are file. However, nobody will check the console if the value and the expected value are the same after changes. Could you instead use some simple assert function to give an error if the received and expected value are different? And no need to console.log the expected value. In my mind this function serves two purposes: the basic tests; but also minimal documentation on how to use the functions.

Bud-Macaulay committed:
rename pt to periodic_table, remove unused switches, console.log -> console.assert

Bud-Macaulay committed:
clarified print-out

Bud-Macaulay commented:
Decided to remove these switches; i cant forsee a use case at the moment.

The assert checks uses a 'tolerance' to determine if the result is close to the hand-coded expected values.

eimrek reviewed (approved):
looks great! thanks a lot! :)
